packages/edit-site/src/components/global-styles/global-styles-provider.js

```diff
@@ -30,6 +33,59 @@ export function mergeBaseAndUserConfigs( base, user ) {
 	} );
 }
 
+// Resolves shared block style variation definitions from the user origin
+// under their respective block types and registers the block style if required.
+function useResolvedBlockStyleVariationsConfig( userConfig ) {
+	const getBlockStyles = useSelect( ( select ) => {
+		return select( blocksStore ).getBlockStyles;
+	}, [] );
+
+	return useMemo( () => {
+		const sharedVariations = userConfig?.styles?.blocks?.variations;
+
+		if ( ! sharedVariations ) {
+			return userConfig;
+		}
+
+		const variationsConfig = cloneDeep( userConfig );
+
+		Object.entries( sharedVariations ).forEach(
+			( [ variationName, variation ] ) => {
+				if ( ! variation?.supportedBlockTypes ) {
+					return;
+				}
+
+				variation.supportedBlockTypes.forEach( ( blockName ) => {
+					// Register any block style variations that have been added
+					// by a theme style variation and are not already registered.
+					const blockStyles = getBlockStyles( blockName );
+					const registeredBlockStyle = blockStyles.find(
+						( { name } ) => name === variationName
+					);
+
+					if ( ! registeredBlockStyle ) {
+						registerBlockStyle( blockName, {
+							name: variationName,
+							label: variationName,
+						} );
```

Comment:
I could split this out into a `useEffect` with `userConfig?.styles?.blocks?.variations` as its dependency. 

As for alternatives on how I could register these block styles outside of when the global styles data is being prepared in the site editor, I'm at a bit of a loss.

Reply:
@talldan let me know if you think https://github.com/WordPress/gutenberg/pull/57908/commits/7a8301154cb13455997532b378f8bfdc0347e68c is any better of if the whole thing needs a rethink. 

Very much open to any suggestions 🙏

Reply:
Yeah, it's a tricky one. Possibly this could be treated more like a pipeline of tasks, so the first task could be a `useMemo` that calculates the list of `unregisteredBlockStyleVariations`, then the other two tasks to carry out based on that list is to:
- register the variation
- merge the variation into the `userConfig`.

It might be possible to do those things in one `useEffect` 🤔 

Alternatively, it might be possible to calculate the subset of data that needs to be merged into `userConfig` first, and then use that data to register the styles and then finally perform the deep merge.

I'd be interested in @kevin940726's thoughts.

I'm also thinking that the approach of cloning the `userConfig` data, setting values in the cloned data, and then merging back into the `userConfig` (which ultimately creates another clone) is possibly too much. Could the `setNestedValue` code instead set values into an empty object, which then gets deepMerged into `userConfig`?

Reply:
I'm a bit lost as to the "why" behind some of the suggestions here, so take this with a grain of salt.

> I'm also thinking that the approach of cloning the userConfig data, setting values in the cloned data, and then merging back into the userConfig (which ultimately creates another clone) is possibly too much.

It's the `userConfig` that needs to be merged back over the resolved variation data so that user made tweaks are maintained.

>  Could the setNestedValue code instead set values into an empty object, which then gets deepMerged into userConfig?

From memory, which could be wrong, I needed the cloned data so everything was still present when I merged the user config back over the data with the resolved variations under block types.

At first glance, the suggestion makes sense, so I'll have to go back make it an empty object again and check the results.

Reply:
> I'm a bit lost as to the "why" behind some of the suggestions here, so take this with a grain of salt.

The main point originally was to avoid the side effect in `useMemo`, which is now solved.

My other comment (before you made the fix) was more about finding ways to fix that, but in a way that didn't introduce adding another loop through the same data again. Easier to explain what I meant through pseudocode:

```js
const sharedVariations = userConfig?.styles?.blocks?.variations;
const [ userConfigToMerge, unregisteredStyles ] = useMemo( () => {
    // logic in here is roughly the same as what you had originally, but instead of calling `registerBlockStyle` you
    // instead push the unregistered styles onto an array that's returned by the useMemo.
    // Also it might be able to avoid the cloning of `userConfig` and instead `setdeep` values into an empty object.
}, [ sharedVariations, hasRegisteredBlockStyle ] );

// This only performs any work when there are unregisted styles.
useEffect(
    () => unregisteredStyles.forEach( ( unregisteredStyle ) => registerBlockStyle( unregisteredStyle ), 
    [ unregisteredStyles ]
); 

// This only updates when `userConfig` changes, which is another optimization as there may be the case where `userConfig` could change, but `userConfig?.styles?.blocks?.variations` is unchanged (referentially).
const updatedConfig = useMemo(
    () => deepMerge( userConfigToMerge, userConfig ),
    [ userConfig ]
);
 
 return updatedConfig;
 ```

Reply:
I've updated the global styles provider in https://github.com/WordPress/gutenberg/pull/57908/commits/78d2a0366d84aaa155c71e39f180996c6be4d493 in  line with the suggested optimisation here.

Once I finally got the performance tests running locally across both trunk and this PR branch, the results were a bit inconclusive given the small number of rounds I could run the tests for.

That said, the code is cleaner, so I've pushed the commit.